### PR TITLE
Speedup ContextInstances if there are null instances

### PR DESCRIPTION
**Pull request by franz1981**
This is an alternative to #38737 with no need of cutoff values and always performing fast-path unlocked checks, if possible, and lazy allocation of a `ReentrantLock`.

In order to reduce the stack-map it has introduced smaller methods to remove/lazy allocate locks which are reused whenever possible

### CODE REVIEW

**Comment by franz1981** [comment]
@Ladicek @mkouba

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java [hunk 1]
```diff
@@ -75,105 +78,128 @@ Collection<Resource> generate(DotName scope) {
...
+            lazyLockMethods.put(namedFields.getKey(), lazyLockMethod.getMethodDescriptor());
+        }
+        return lazyLockMethods;
+    }
+
     private void implementGetAllPresent(ClassCreator contextInstances, Map<String, InstanceAndLock> idToFields) {
         MethodCreator getAllPresent = contextInstances.getMethodCreator("getAllPresent", Set.class)
                 .setModifiers(ACC_PUBLIC);
-        // this.lockAll();
         // ContextInstanceHandle<?> copy1 = this.1;
```
Comment (franz1981) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java:171]: I've removed the lockAll/unlockAll here, since it doesn't appear for them to be useful @Ladicek @mkouba

**Comment by franz1981** [comment]
@mkouba I've added 92bc2e542a8f2d316bff0f71785f327ff9cd199a and update the Java code snippet to explain what is doing: it shouldn't be a big deal but it seems more correct to return only the value read, instead of performing an unguarded volatile read, there.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java [hunk 2]
```diff
@@ -66,6 +68,7 @@ Collection<Resource> generate(DotName scope) {
         // Add ContextInstanceHandle and Lock fields for every bean
         // The name of the field is a generated index
         // For example:
+        // private static final AtomicReferenceFieldUpdater<ContextInstances, ContextInstanceHandle> LAZY_1L_UPDATER;
         // private volatile ContextInstanceHandle 1;
         // private final Lock 1l = new ReentrantLock();
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java:73]: The lock field is now `volatile` as well, right?

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java (hunk 1 above)
Comment (mkouba) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java:171]: So the original idea was to get the snapshot at the given time, i.e. disallow writing during the invocation of this method but you're right that it's not very useful. ​The truth is also that this method is almost never used. And so +1 ;-).

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java [hunk 3]
```diff
@@ -75,105 +78,128 @@ Collection<Resource> generate(DotName scope) {
...
+    private void implementRemoveEach(ClassCreator contextInstances, List<MethodDescriptor> removeInstances) {
         MethodCreator removeEach = contextInstances.getMethodCreator("removeEach", void.class, Consumer.class)
                 .setModifiers(ACC_PUBLIC);
-        // this.lockAll();
-        // ContextInstanceHandle<?> copy1 = this.1;
-        // if (copy1 != null) {
-        //    this.1 = null;
-        // }
-        // this.unlockAll();
+        // ContextInstanceHandle<?> copy1 = remove1();
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java:191]: Hm, so if we remove `lockAll()` then this operation is not atomic (in the sense of the set of context instances we operate on), right? And that's probably ok, because it's only used during the destruction of a context.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java [hunk 4]
```diff
@@ -182,42 +208,49 @@ private void implementRemoveEach(ClassCreator contextInstances, Map<String, Inst
...
             // ContextInstanceHandle<?> copy = this.1;
-            // if (copy != null) {
-            //    this.1 = null;
+            // if (copy == null) {
+            //  return null;
             // }
-            // this.1l.unlock();
+            // Lock lock = lazy1l();
+            // lock.lock();
+            // copy = this.1;
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java:234]: I think that we should either skip the read operation `copy = this.1;` or replace `this.1 = null;` with `if (copy != null) { this.1 = null; }` since value could be different after the lock is acquired...?

**Review by mkouba — COMMENTED** [review]
I've added few comments but it looks really good!

I will try to run our microbenchmarks to see if it makes any difference...

**Comment by franz1981** [comment]
Thanks @mkouba yep, in case the micro doesn't impact, I suggest to perform one where the context instance is allocated on the fly and have different instance fields while using a parameter to decide how many to compute, and including a remove at the end of the usage life-cycle ; having such benchmark will help capturing the actual possible behaviours for request scoped beans, which seems to be allocated, computed, removed in the hot path

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java (hunk 4 above)
Comment (mkouba) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java:234]: @franz1981 and this one?

**Comment by mkouba** [comment]
So I ran the microbenchmarks and the only benchmark that shows _significant_ improvement is the [`RequestContextActivationBenchmark`](https://github.com/mkouba/arc-benchmarks/blob/master/src/main/java/io/quarkus/arc/benchmarks/RequestContextActivationBenchmark.java) which merely tests request context activation/deactivation, no bean operations - and it's expected due to lazy init of `Lock`. The throughtput is doubled! 

The [`ContextProviderBenchmark`](https://github.com/mkouba/arc-benchmarks/blob/master/src/main/java/io/quarkus/arc/benchmarks/ContextProviderBenchmark.java) that tests `ArcContextProvider` (SmallRye Context Propagation integration) shows a decent improvement too (because it does not do much more than `RequestContextActivationBenchmark`).  So this PR could slightly improve the performance in use cases where SR Context Propagation is used extensively.

However, the [RequestContextBenchmark](https://github.com/mkouba/arc-benchmarks/blob/master/src/main/java/io/quarkus/arc/benchmarks/RequestContextBenchmark.java), which activates the context, invokes client proxies of 5 `@RequestScoped` beans and finally terminates the context, shows no significant change. And I think that it's expected too because this PR does not improve the `ContextInstances#computeIfAbsent()` method which is critical for this benchmark. @franz1981 is the use case you describe in the previous comment different?

**Comment by franz1981** [comment]
> @franz1981 is the use case you describe in the previous comment different?

Exactly and happy that you already prepared other benchmarks which already capture the motivation of the changes sent here, thanks again for checking!!!!

**Comment by mkouba** [comment]
> Probably [RequestContextBenchmark](https://github.com/mkouba/arc-benchmarks/blob/master/src/main/java/io/quarkus/arc/benchmarks/RequestContextBenchmark.java), could have been relevant if the benchmark had a parameter to decide how many of these beans are computed eg for a total of 5 beans, just 1(,2,3,4,5) is computed. It will impact the removal as well, given that this pr remove only what is actually computed, too.

So I've tried to adapt the `RequestContextBenchmark` in a way that:
1. The app contains 10 request scoped beans
1. We always test 15 client proxy invocations but in three variants:
   1) 1 of the 10 beans is used (i.e. instantiated in `ContextInstances.computeIfAbsent()`), 
   2) 3 of the 10 beans are used, and 
   3) 5 of the 10 beans are used. 

And this PR indeed increased the throughput in the first scenario (1/10) by ~ 40%, in the second (3/10) by ~ 15% and the results of the third are more or less the same. 

In other words, the more request scoped beans exist in the app and the less beans are actually used in the benchmark the better throughput with this PR. And since it's very likely that only a fraction of all request scoped beans is used in a single request of a "real world" app I think that it's really a great improvement. Good job!

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java (hunk 4 above)
Comment (franz1981) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java:234]: IDK, really...
the case in which this would happen can only be if:
- 2 removes are concurrently acquiring the lock after read a non-null instance: we do expect the second one to NOT perform any nulling out while within the lock, because the value has been already removed. Performing it again is a waste, but it isn't "wrong" let's say.
- 1 remove read a non-null instance, another remove cleanup everything and a compute acquire the lock, allocating a new instance: if that happen, the remove will enter and observe a different value from the first one, but is still fine if it read and remove it, I think. The problem is that it loose linearizability (but is due to the unguarded volatile load, as said in other comments).

In short, performing volatile reads always bring a new state in the mix, which makes all the operations non-atomic(s); we could do something must more "precise" here by creating an holder class `InstanceAndLock`, as said in other comments. This can make a much better choice here.

Has to be said, that, at least of request scoped beans we have an earlier atomic (based on compare and set) invalidation, which fully avoid remove to happen twice on the same context, which is the key requirement to make the latter case to be triggered.

**Comment by franz1981** [comment]
I'm adding a note here to help ourselves of the future: in order to "fix" the existing impl and this PR in the case where a compute started after an invalidation can "revive" and reallocate a bean:
1. compute always have to check for any existing invalidation *before* starting
2. after compute has happened, in case it has stored any value, it should check again the invalidation status, and if `true`, should loop till removing `null` out everything

The last loop is necessary because other(s) concurrent compute can end up "recomputing" the value again, so it's important for them to keep on trying to remove the value till is stable as `null`.

This mechanism is to prevent `@PreDestroy` to NOT be called because the remove will read a `null` volatile instance value, and the compute will set it right after, without cleaning it up.

@mkouba @Ladicek 
I know that's crazy but i can easily forget this concurrent stuff, so let me brain dump this here

**Comment by franz1981** [comment]
Ok @geoand this should be a very good one, once merged. Just need to make sure the test failures are all flaky ones

**Comment by franz1981** [comment]
https://github.com/franz1981/quarkus/actions/runs/7941041918 seems to show that the failure reported here is real or another flaky test

**Comment by geoand** [comment]
@mkouba this looks good, WDYT?

**Comment by franz1981** [comment]
Done @mkouba and @geoand waiting the CI again, although I've just changed a comment and rebased)

**Review by mkouba — APPROVED** [review]

**Review by Ladicek — APPROVED** [review]
`getAllPresent()` will no longer return a consistent snapshot, but that's not guaranteed for `ComputingCacheContextInstances` either, so should be fine. Nice work!
